android/src/app/organicmaps/location/SensorHelper.java

```diff
@@ -5,67 +5,101 @@
   {
-    if (!mMwmApplication.arePlatformAndCoreInitialized())
-      return;
+    // Here we can have events from one out of these two sensors:
+    // TYPE_GEOMAGNETIC_ROTATION_VECTOR
+    // TYPE_ROTATION_VECTOR
 
-    notifyInternal(event);
-  }
+    if (mLastAccuracy != event.accuracy)
+    {
+      mLastAccuracy = event.accuracy;
+      String toastText = null;
+      switch (mLastAccuracy)
+      {
+        case SensorManager.SENSOR_STATUS_ACCURACY_HIGH:
+          break;
+        case SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM:
+          toastText = mMwmApplication.getString(R.string.compass_calibration_recommended);
+          break;
+        case SensorManager.SENSOR_STATUS_ACCURACY_LOW:
+        case SensorManager.SENSOR_STATUS_UNRELIABLE:
+        default:
+          toastText = mMwmApplication.getString(R.string.compass_calibration_required);
+      }
+      if (toastText != null)
+        Toast.makeText(mMwmApplication, toastText, Toast.LENGTH_LONG).show();
+    }
 
-  private void notifyInternal(@NonNull SensorEvent event)
-  {
-    if (event.sensor.getType() == Sensor.TYPE_ROTATION_VECTOR) {
-      float[] rotMatrix = new float[9];
-      SensorManager.getRotationMatrixFromVector(rotMatrix, event.values);
-      SensorManager.remapCoordinateSystem(rotMatrix,
-                                          SensorManager.AXIS_X, SensorManager.AXIS_Y, rotMatrix);
+    if (event.accuracy == SensorManager.SENSOR_STATUS_UNRELIABLE)
+      return;
 
-      float[] rotVals = new float[3];
-      SensorManager.getOrientation(rotMatrix, rotVals);
+    SensorManager.getRotationMatrixFromVector(mRotationMatrix, event.values);
+    SensorManager.remapCoordinateSystem(mRotationMatrix, SensorManager.AXIS_X, SensorManager.AXIS_Y, mRotationMatrix);
 
-      // rotVals indexes: 0 - yaw, 2 - roll, 1 - pitch.
-      LocationHelper.INSTANCE.notifyCompassUpdated(rotVals[0]);
-    }
+    SensorManager.getOrientation(mRotationMatrix, mRotationValues);
+
+    // mRotationValues indexes: 0 - yaw, 2 - roll, 1 - pitch.
+    LocationHelper.INSTANCE.notifyCompassUpdated(mRotationValues[0]);
   }
 
   @Override
-  public void onAccuracyChanged(Sensor sensor, int accuracy) {}
+  public void onAccuracyChanged(Sensor sensor, int accuracy)
+  {
+    Log.w("onAccuracyChanged", "Sensor " + sensor.getStringType() + " has changed accuracy to " + accuracy);
+    // This method is called _only_ when accuracy changes. To know the initial startup accuracy,
+    // and to show calibration warning toast if necessary, we check it in onSensorChanged().
+  }
 
   SensorHelper(@NonNull Context context)
   {
     mMwmApplication = MwmApplication.from(context);
     mSensorManager = (SensorManager) mMwmApplication.getSystemService(Context.SENSOR_SERVICE);
-    if (mSensorManager != null)
-      mRotation = mSensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
+    Sensor sensor = mSensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
```

Comment:
Was ```if (mSensorManager != null)``` check before. Now its gone, ok?

Reply:
I don't think that there are some android devices without physical or software sensors.

Reply:
Meaning, this API should always return non-null.